File: Programmers/chaeeun/디스크컨트롤러.py
# (참고)
# 최소 힙

import heapq

def solution(jobs):
    answer, now, i = 0, 0, 0
    start = -1
    heap = []

    while i < len(jobs):
        for j in jobs:
            if start < j[0] <= now:
                heapq.heappush(heap, [j[1], j[0]])
                
        if len(heap) > 0:
            current = heapq.heappop(heap)
            start = now
            now += current[0]
            answer += (now - current[1])
            i += 1
            
        else: # 남아있는 작업들의 요청시간이 아직 오지 않음
            now += 1

    # print(int(answer / len(jobs)))
    return int(answer / len(jobs))

# solution([[0, 3], [1, 9], [2, 6]])

File: Programmers/myeongju/디스크컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        
        PriorityQueue<Job> queue = new PriorityQueue<>();
        PriorityQueue<Job> list = new PriorityQueue<>(new Comparator<Job>() {
            @Override
            public int compare(Job o1, Job o2) {
                return o1.work - o2.work;
            }
        });

        for(int[] arr : jobs)
            queue.add(new Job(arr[0],arr[1]));
        
        int time = 0;
        while(true) {
            
            if(list.isEmpty()) {
                if(queue.isEmpty()) break;
                else list.add(queue.poll());
            }
            
            Job now = list.poll();
            
            // 현재 작업이 시작하기 까지 대기한 시간을 더해준다.
            if(time > now.start)
                answer += time - now.start;
            else time = now.start;
            
            // 작업 시간을 더해준다.
            time += now.work;
            answer += now.work;
            
            while(!queue.isEmpty() && queue.peek().start <= time)
                list.add(queue.poll());
        }
        
        return answer/jobs.length;
    }
    
    public class Job implements Comparable<Job> {
        int start;
        int work;
        
        public Job(int start, int work) {
            this.start = start;
            this.work = work;
        }
        
        // 요청 시간 오름차순 -> 작업 시간 오름차순
        @Override
        public int compareTo(Job o) {
            if(this.start==o.start) 
                return this.work-o.work;
            else return this.start-o.start;
        }
    }
}